### PR TITLE
8303018: Unicode Emoji Properties

### PR DESCRIPTION
**Pull request by naotoj**
Proposing accessor methods to Emoji properties defined in [Unicode Technical Standard #51](https://unicode.org/reports/tr51/) in `java.lang.Character` class. This is per a request from the client group, as well as refining the currently existing ad-hoc emoji implementation in regex. A CSR has also been drafted, and I would appreciate reviews for it too.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8304027](https://bugs.openjdk.org/browse/JDK-8304027) to be approved

### Issues
 * [JDK-8303018](https://bugs.openjdk.org/browse/JDK-8303018): Unicode Emoji Properties
 * [JDK-8304027](https://bugs.openjdk.org/browse/JDK-8304027): Unicode Emoji Properties (**CSR**)

### Reviewers
 * [Phil Race](https://openjdk.org/census#prr) (@prrace - **Reviewer**) ⚠️ Review applies to [4b2a2418](https://git.openjdk.org/jdk/pull/13006/files/4b2a241861061cd21e48f6ad4a999be4d21dd1bd)
 * [Erik Joelsson](https://openjdk.org/census#erikj) (@erikj79 - **Reviewer**) ⚠️ Review applies to [4b2a2418](https://git.openjdk.org/jdk/pull/13006/files/4b2a241861061cd21e48f6ad4a999be4d21dd1bd)
 * [Roger Riggs](https://openjdk.org/census#rriggs) (@RogerRiggs - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13006/head:pull/13006` \
`$ git checkout pull/13006`

Update a local copy of the PR: \
`$ git checkout pull/13006` \
`$ git pull https://git.openjdk.org/jdk.git pull/13006/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13006`

View PR using the GUI difftool: \
`$ git pr show -t 13006`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13006.diff">https://git.openjdk.org/jdk/pull/13006.diff</a>

</details>

### CODE REVIEW

**Review by prrace — APPROVED** [review]
Not at all familiar with all these templates but everything I do understand lgtm :-)

File: src/java.base/share/classes/java/lang/Character.java [hunk 1]
```diff
@@ -10781,6 +10781,113 @@ public static boolean isIdentifierIgnorable(int codePoint) {
...
+     * the {@code Extended_Pictographic} property, defined in
+     * <a href="https://unicode.org/reports/tr51/#Emoji_Properties_and_Data_Files">
+     * Unicode Emoji (Technical Standard #51)</a>.
+     *
+     * @param   codePoint the character (Unicode code point) to be tested.
+     * @return  {@code true} if the character is an Extended Pictographic;
+     *          {@code false} otherwise.
+     * @since   21
+     */
+    public static boolean isExtendedPictographic(int codePoint) {
```
Comment (AlanBateman) [inline src/java.base/share/classes/java/lang/Character.java:10887]: Is there a "cut 'n paste" error here, it looks like the description has been copied from  isEmojiComponent.

File: src/java.base/share/classes/java/lang/Character.java [hunk 2]
```diff
@@ -10781,6 +10781,113 @@ public static boolean isIdentifierIgnorable(int codePoint) {
...
+     *          {@code false} otherwise.
+     * @since   21
+     */
+    public static boolean isEmoji(int codePoint) {
+        return CharacterData.of(codePoint).isEmoji(codePoint);
+    }
+
+    /**
+     * Determines if the specified character (Unicode code point) has the
+     * Emoji Presentation by default.
```
Comment (AlanBateman) [inline src/java.base/share/classes/java/lang/Character.java:10803]: "has the Emoji Presentation", should that be "has the Emoji Presentation property"?

**Review by erikj79 — APPROVED** [review]
Build changes look ok.

File: src/java.base/share/classes/java/lang/Character.java (hunk 2 above)
Comment (naotoj) [reply src/java.base/share/classes/java/lang/Character.java]: Fixed this one as well

File: src/java.base/share/classes/java/lang/Character.java (hunk 1 above)
Comment (naotoj) [reply src/java.base/share/classes/java/lang/Character.java:10887]: My bad. You are right. Corrected.

File: make/jdk/src/classes/build/tools/generatecharacter/GenerateCharacter.java [hunk 3]
```diff
@@ -188,15 +197,21 @@ the values in the table can be preshifted (generally possible if the table
...
+        maskOtherAlphabetic = 0x000400000000L,
+        maskIdeographic     = 0x000800000000L,
+        maskIDStart         = 0x001000000000L,
+        maskIDContinue      = 0x002000000000L,
+        maskEmoji           = 0x004000000000L,
+        maskEmojiPresentation = 0x008000000000L,
+        maskEmojiModifier   = 0x010000000000L,
+        maskEmojiModifierBase = 0x020000000000L,
+        maskEmojiComponent  = 0x040000000000L,
+        maskExtendedPictographic = 0x080000000000L;
```
Comment (eirbjo) [inline make/jdk/src/classes/build/tools/generatecharacter/GenerateCharacter.java:215]: Would be nice the have the '=' aligned in the same column as the ones above, or at least have the emoji ones align with each other.

**Comment by eirbjo** [comment]
Would it make sense to add something like the following to `GenerateCharacter.propertiesComments`?

```java
if ((val & maskEmoji) == maskEmoji) {
    result.append(", emoji ");
}
if ((val & maskEmojiPresentation) == maskEmojiPresentation) {
    result.append(", emojiPresentation ");
}
if ((val & maskEmojiModifier) == maskEmojiModifier) {
    result.append(", emojiModifier ");
}
if ((val & maskEmojiModifierBase) == maskEmojiModifierBase) {
    result.append(", emojiModifierBase ");
}
if ((val & maskEmojiComponent) == maskEmojiComponent) {
    result.append(", emojiComponent ");
}
if ((val & maskExtendedPictographic) == maskExtendedPictographic) {
    result.append(", extendedPictographic ");
}
```

This updates the comments of the B table to something like this sample:

```
0x0000,  //  34   unassigned, L
0x0440,  //  35   unassigned, L, emoji , emojiComponent
0x0000,  //  36   unassigned, L
```

File: make/jdk/src/classes/build/tools/generatecharacter/GenerateCharacter.java [hunk 4]
```diff
@@ -776,6 +802,12 @@ static String replaceCommand(String x) {
         if (x.equals("maskIdeographic")) return "0x" + hex4(maskIdeographic >> 32);
         if (x.equals("maskIDStart")) return "0x" + hex4(maskIDStart >> 32);
         if (x.equals("maskIDContinue")) return "0x" + hex4(maskIDContinue >> 32);
+        if (x.equals("maskEmoji")) return "0x" + hex4(maskEmoji >> 32);
+        if (x.equals("maskEmojiPresentation")) return "0x" + hex4(maskEmojiPresentation >> 32);
+        if (x.equals("maskEmojiModifier")) return "0x" + hex4(maskEmojiModifier >> 32);
+        if (x.equals("maskEmojiModifierBase")) return "0x" + hex4(maskEmojiModifierBase >> 32);
+        if (x.equals("maskEmojiComponent")) return "0x" + hex4(maskEmojiComponent >> 32);
+        if (x.equals("maskExtendedPictographic")) return "0x" + hex4(maskExtendedPictographic >> 32);
```
Comment (RogerRiggs) [inline make/jdk/src/classes/build/tools/generatecharacter/GenerateCharacter.java:810]: An upgrade would be to modify hex4(), hexNN() to use `HexFormat.of().toUpperCase().toHexDigits((short)xxx)`
The HexFormat is reusable and would avoid creating extra strings.
Perhaps also create a method that combines the repetitive shift and prefixing.

This if...then... sequence could be an expression switch (x) {...}.

**Review by RogerRiggs — APPROVED** [review]
Looks good.  
There are opportunities to modernize the code, but maybe separately.

File: make/jdk/src/classes/build/tools/generatecharacter/GenerateCharacter.java (hunk 3 above)
Comment (naotoj) [reply make/jdk/src/classes/build/tools/generatecharacter/GenerateCharacter.java]: Thanks. Indentation aligned, and although in production `-string` is used so comments are not emitted, added information for `B` table in `GenerateCharacter.propertiesComments` for consistency.

File: make/jdk/src/classes/build/tools/generatecharacter/EmojiData.java [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+
+    private static long convertType(String type) {
+        return switch (type) {
+            case "Emoji" -> EMOJI;
+            case "Emoji_Presentation" -> EMOJI_PRESENTATION;
+            case "Emoji_Modifier" -> EMOJI_MODIFIER;
+            case "Emoji_Modifier_Base" -> EMOJI_MODIFIER_BASE;
+            case "Emoji_Component" -> EMOJI_COMPONENT;
+            case "Extended_Pictographic" -> EXTENDED_PICTOGRAPHIC;
+            default -> throw new InternalError();
```
[the remark this answers is not in the document]
Comment (naotoj) [reply make/jdk/src/classes/build/tools/generatecharacter/EmojiData.java]: Added `type` to the error message

File: make/jdk/src/classes/build/tools/generatecharacter/GenerateCharacter.java [hunk 6]
```diff
@@ -188,15 +197,21 @@ the values in the table can be preshifted (generally possible if the table
...
+        maskOtherAlphabetic = 0x000400000000L,
+        maskIdeographic     = 0x000800000000L,
+        maskIDStart         = 0x001000000000L,
+        maskIDContinue      = 0x002000000000L,
+        maskEmoji           = 0x004000000000L,
+        maskEmojiPresentation = 0x008000000000L,
+        maskEmojiModifier   = 0x010000000000L,
+        maskEmojiModifierBase = 0x020000000000L,
+        maskEmojiComponent  = 0x040000000000L,
+        maskExtendedPictographic = 0x080000000000L;
```
[the remark this answers is not in the document]
Comment (naotoj) [reply make/jdk/src/classes/build/tools/generatecharacter/GenerateCharacter.java]: Good point. I managed to get rid of the constants in `EmojiData` altogether, by using the constants in `GenerateCharacter`. Used the bit numbers to construct constants.

File: make/jdk/src/classes/build/tools/generatecharacter/GenerateCharacter.java (hunk 4 above)
Comment (naotoj) [reply make/jdk/src/classes/build/tools/generatecharacter/GenerateCharacter.java:810]: It would be good, but it would be for another day IMO.

**Review by RogerRiggs — APPROVED** [review]

**Comment by eirbjo** [comment]
Unrelated side note: Reviewing this PR inspired me to see if one could generate more efficient switch expressions for the CharacterDataLatin1 methods as replacement for the property lookup / masking done today. This seemed to give a small improvements on benchmarks and also collapsed a few methods to simply "return false". Could be something to explore at a later point.

**Comment by naotoj** [comment]
> Unrelated side note: Reviewing this PR inspired me to see if one could generate more efficient switch expressions for the CharacterDataLatin1 methods as replacement for the property lookup / masking done today. This seemed to give a small improvements on benchmarks and also collapsed a few methods to simply "return false". Could be something to explore at a later point.

Yeah, agree. `Latin1` can be (should be?) fast-path'ed as always.

**Pull request closed without merging**